File: android-app/app/src/main/java/com/chainlesschain/android/remote/commands/InputCommands.kt
package com.chainlesschain.android.remote.commands

import com.chainlesschain.android.remote.client.RemoteCommandClient
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 输入控制命令 API
 *
 * 提供类型安全的键盘和鼠标控制命令
 */
@Singleton
class InputCommands @Inject constructor(
    private val client: RemoteCommandClient
) {
    /**
     * 发送单个按键
     *
     * @param key 按键名称（如 enter, tab, escape, f1-f12, a-z, 0-9 等）
     */
    suspend fun sendKeyPress(key: String): Result<KeyPressResponse> {
        val params = mapOf("key" to key)
        return client.invoke("input.sendKeyPress", params)
    }

    /**
     * 发送组合键
     *
     * @param key 主按键
     * @param modifiers 修饰键列表（ctrl, alt, shift, meta, win, cmd）
     */
    suspend fun sendKeyCombo(
        key: String,
        modifiers: List<String> = emptyList()
    ): Result<KeyComboResponse> {
        val params = mapOf(
            "key" to key,
            "modifiers" to modifiers
        )
        return client.invoke("input.sendKeyCombo", params)
    }

    /**
     * 输入文本
     *
     * @param text 要输入的文本
     * @param delay 字符间延迟（毫秒）
     */
    suspend fun typeText(
        text: String,
        delay: Int = 50
    ): Result<TypeTextResponse> {
        val params = mapOf(
            "text" to text,
            "delay" to delay
        )
        return client.invoke("input.typeText", params)
    }

    /**
     * 移动鼠标
     *
     * @param x X 坐标
     * @param y Y 坐标
     * @param relative 是否相对移动
     */
    suspend fun mouseMove(
        x: Int,
        y: Int,
        relative: Boolean = false
    ): Result<MouseMoveResponse> {
        val params = mapOf(
            "x" to x,
            "y" to y,
            "relative" to relative
        )
        return client.invoke("input.mouseMove", params)
    }

    /**
     * 鼠标点击
     *
     * @param button 按钮：left, right, middle
     * @param x X 坐标（可选，如提供则先移动到该位置）
     * @param y Y 坐标（可选）
     */
    suspend fun mouseClick(
        button: String = "left",
        x: Int? = null,
        y: Int? = null
    ): Result<MouseClickResponse> {
        val params = mutableMapOf<String, Any>("button" to button)
        x?.let { params["x"] = it }
        y?.let { params["y"] = it }

        return client.invoke("input.mouseClick", params)
    }

    /**
     * 鼠标双击
     *
     * @param button 按钮：left, right, middle
     * @param x X 坐标（可选）
     * @param y Y 坐标（可选）
     */
    suspend fun mouseDoubleClick(
        button: String = "left",
        x: Int? = null,
        y: Int? = null
    ): Result<MouseClickResponse> {
        val params = mutableMapOf<String, Any>("button" to button)
        x?.let { params["x"] = it }
        y?.let { params["y"] = it }

        return client.invoke("input.mouseDoubleClick", params)
    }

    /**
     * 鼠标拖拽
     *
     * @param startX 起始 X 坐标
     * @param startY 起始 Y 坐标
     * @param endX 结束 X 坐标
     * @param endY 结束 Y 坐标
     * @param button 按钮：left, right, middle
     */
    suspend fun mouseDrag(
        startX: Int,
        startY: Int,
        endX: Int,
        endY: Int,
        button: String = "left"
    ): Result<MouseDragResponse> {
        val params = mapOf(
            "startX" to startX,
            "startY" to startY,
            "endX" to endX,
            "endY" to endY,
            "button" to button
        )
        return client.invoke("input.mouseDrag", params)
    }

    /**
     * 鼠标滚动
     *
     * @param direction 方向：up, down, left, right
     * @param amount 滚动量
     */
    suspend fun mouseScroll(
        direction: String = "down",
        amount: Int = 3
    ): Result<MouseScrollResponse> {
        val params = mapOf(
            "direction" to direction,
            "amount" to amount
        )
        return client.invoke("input.mouseScroll", params)
    }

    /**
     * 获取鼠标位置
     */
    suspend fun getCursorPosition(): Result<InputCursorPositionResponse> {
        return client.invoke("input.getCursorPosition", emptyMap())
    }

    /**
     * 获取键盘布局
     */
    suspend fun getKeyboardLayout(): Result<KeyboardLayoutResponse> {
        return client.invoke("input.getKeyboardLayout", emptyMap())
    }

    // 便捷方法

    /**
     * 发送 Enter 键
     */
    suspend fun pressEnter(): Result<KeyPressResponse> = sendKeyPress("enter")

    /**
     * 发送 Escape 键
     */
    suspend fun pressEscape(): Result<KeyPressResponse> = sendKeyPress("escape")

    /**
     * 发送 Tab 键
     */
    suspend fun pressTab(): Result<KeyPressResponse> = sendKeyPress("tab")

    /**
     * 发送 Backspace 键
     */
    suspend fun pressBackspace(): Result<KeyPressResponse> = sendKeyPress("backspace")

    /**
     * 发送 Ctrl+C（复制）
     */
    suspend fun copy(): Result<KeyComboResponse> = sendKeyCombo("c", listOf("ctrl"))

    /**
     * 发送 Ctrl+V（粘贴）
     */
    suspend fun paste(): Result<KeyComboResponse> = sendKeyCombo("v", listOf("ctrl"))

    /**
     * 发送 Ctrl+X（剪切）
     */
    suspend fun cut(): Result<KeyComboResponse> = sendKeyCombo("x", listOf("ctrl"))

    /**
     * 发送 Ctrl+Z（撤销）
     */
    suspend fun undo(): Result<KeyComboResponse> = sendKeyCombo("z", listOf("ctrl"))

    /**
     * 发送 Ctrl+A（全选）
     */
    suspend fun selectAll(): Result<KeyComboResponse> = sendKeyCombo("a", listOf("ctrl"))

    /**
     * 发送 Ctrl+S（保存）
     */
    suspend fun save(): Result<KeyComboResponse> = sendKeyCombo("s", listOf("ctrl"))

    /**
     * 左键单击
     */
    suspend fun leftClick(x: Int? = null, y: Int? = null): Result<MouseClickResponse> =
        mouseClick("left", x, y)

    /**
     * 右键单击
     */
    suspend fun rightClick(x: Int? = null, y: Int? = null): Result<MouseClickResponse> =
        mouseClick("right", x, y)

    /**
     * 向上滚动
     */
    suspend fun scrollUp(amount: Int = 3): Result<MouseScrollResponse> =
        mouseScroll("up", amount)

    /**
     * 向下滚动
     */
    suspend fun scrollDown(amount: Int = 3): Result<MouseScrollResponse> =
        mouseScroll("down", amount)
}

// 响应数据类

@Serializable
data class KeyPressResponse(
    val success: Boolean,
    val key: String,
    val message: String
)

@Serializable
data class KeyComboResponse(
    val success: Boolean,
    val modifiers: List<String>,
    val key: String,
    val message: String
)

@Serializable
data class TypeTextResponse(
    val success: Boolean,
    val length: Int,
    val delay: Int,
    val message: String
)

@Serializable
data class MouseMoveResponse(
    val success: Boolean,
    val x: Int,
    val y: Int,
    val relative: Boolean,
    val message: String
)

@Serializable
data class MouseClickResponse(
    val success: Boolean,
    val button: String,
    val x: Int? = null,
    val y: Int? = null,
    val message: String
)

@Serializable
data class MouseDragResponse(
    val success: Boolean,
    val startX: Int,
    val startY: Int,
    val endX: Int,
    val endY: Int,
    val message: String
)

@Serializable
data class MouseScrollResponse(
    val success: Boolean,
    val direction: String,
    val amount: Int,
    val message: String
)

@Serializable
data class InputCursorPositionResponse(
    val success: Boolean,
    val x: Int,
    val y: Int
)

@Serializable
data class KeyboardLayoutResponse(
    val success: Boolean,
    val layout: String,
    val platform: String,
    val error: String? = null
)

File: ios-app/Modules/CoreP2P/Sources/RemoteSkills/Input/InputCommands.swift
import Foundation

/// 远程键鼠输入 typed RPC wrapper — Phase 6.1B1。
///
/// 镜像 Android `app/src/main/java/.../remote/commands/InputCommands.kt`。
/// 10 method 全 wired，桌面 `input-handler.js` 与 Android Commands 100% 名称对齐
/// （Coverage doc §1.4 数据：A=10 D=10 ✓=10）。
///
/// **wire 协议**（与桌面 `input-handler.js` 对齐）：
/// - `input.sendKeyPress` params: `{key}` → `{success, key, message}`
/// - `input.sendKeyCombo` params: `{key, modifiers:[]}` → `{success, modifiers, key, message}`
/// - `input.typeText` params: `{text, delay}` → `{success, length, delay, message}`
/// - `input.mouseMove` params: `{x, y, relative}` → `{success, x, y, relative, message}`
/// - `input.mouseClick` / `mouseDoubleClick` params: `{button, x?, y?}` → `{success, button, x?, y?, message}`
/// - `input.mouseDrag` params: `{startX, startY, endX, endY, button}` → drag response
/// - `input.mouseScroll` params: `{direction, amount}` → scroll response
/// - `input.getCursorPosition` params: `{}` → `{success, x, y}`
/// - `input.getKeyboardLayout` params: `{}` → `{success, layout, platform, error?}`
public actor InputCommands {

    private let client: RemoteCommandClient

    public init(client: RemoteCommandClient) {
        self.client = client
    }

    // MARK: - 键盘

    /// 发送单个按键（如 enter / tab / escape / f1-f12 / a-z / 0-9）。
    public func sendKeyPress(
        pcPeerId: String,
        key: String,
        mobileDid: String? = nil
    ) async throws -> KeyPressResponse {
        guard !key.isEmpty else {
            throw RemoteSkillError.invalidArgument("input.sendKeyPress: key empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.sendKeyPress",
            params: ["key": key],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, KeyPressResponse.decode)
    }

    /// 发送组合键（如 Ctrl+C → key="c", modifiers=[.ctrl]）。
    public func sendKeyCombo(
        pcPeerId: String,
        key: String,
        modifiers: [KeyModifier] = [],
        mobileDid: String? = nil
    ) async throws -> KeyComboResponse {
        guard !key.isEmpty else {
            throw RemoteSkillError.invalidArgument("input.sendKeyCombo: key empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.sendKeyCombo",
            params: [
                "key": key,
                "modifiers": modifiers.map { $0.rawValue }
            ],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, KeyComboResponse.decode)
    }

    /// 输入文本（字符间延迟可调，默认 50ms）。
    public func typeText(
        pcPeerId: String,
        text: String,
        delay: Int = 50,
        mobileDid: String? = nil
    ) async throws -> TypeTextResponse {
        guard !text.isEmpty else {
            throw RemoteSkillError.invalidArgument("input.typeText: text empty")
        }
        guard delay >= 0 else {
            throw RemoteSkillError.invalidArgument("input.typeText: delay must be ≥ 0")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.typeText",
            params: ["text": text, "delay": delay],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, TypeTextResponse.decode)
    }

    // MARK: - 鼠标

    /// 移动鼠标（相对 / 绝对）。
    public func mouseMove(
        pcPeerId: String,
        x: Int,
        y: Int,
        relative: Bool = false,
        mobileDid: String? = nil
    ) async throws -> MouseMoveResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.mouseMove",
            params: ["x": x, "y": y, "relative": relative],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, MouseMoveResponse.decode)
    }

    /// 鼠标点击（可选 x/y — 如提供先移动）。
    public func mouseClick(
        pcPeerId: String,
        button: MouseButton = .left,
        x: Int? = nil,
        y: Int? = nil,
        mobileDid: String? = nil
    ) async throws -> MouseClickResponse {
        var params: [String: Any] = ["button": button.rawValue]
        if let x = x { params["x"] = x }
        if let y = y { params["y"] = y }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.mouseClick",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, MouseClickResponse.decode)
    }

    /// 鼠标双击。
    public func mouseDoubleClick(
        pcPeerId: String,
        button: MouseButton = .left,
        x: Int? = nil,
        y: Int? = nil,
        mobileDid: String? = nil
    ) async throws -> MouseClickResponse {
        var params: [String: Any] = ["button": button.rawValue]
        if let x = x { params["x"] = x }
        if let y = y { params["y"] = y }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.mouseDoubleClick",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, MouseClickResponse.decode)
    }

    /// 鼠标拖拽（startX/Y → endX/Y）。桌面端标 ADMIN 权限。
    public func mouseDrag(
        pcPeerId: String,
        startX: Int,
        startY: Int,
        endX: Int,
        endY: Int,
        button: MouseButton = .left,
        mobileDid: String? = nil
    ) async throws -> MouseDragResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.mouseDrag",
            params: [
                "startX": startX, "startY": startY,
                "endX": endX, "endY": endY,
                "button": button.rawValue
            ],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, MouseDragResponse.decode)
    }

    /// 鼠标滚动。
    public func mouseScroll(
        pcPeerId: String,
        direction: ScrollDirection = .down,
        amount: Int = 3,
        mobileDid: String? = nil
    ) async throws -> MouseScrollResponse {
        guard amount > 0 else {
            throw RemoteSkillError.invalidArgument("input.mouseScroll: amount must be > 0")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.mouseScroll",
            params: ["direction": direction.rawValue, "amount": amount],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, MouseScrollResponse.decode)
    }

    // MARK: - 查询

    /// 获取当前鼠标位置（只读）。
    public func getCursorPosition(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> CursorPositionResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.getCursorPosition",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, CursorPositionResponse.decode)
    }

    /// 获取键盘布局（如 US / GB / DE）+ 平台信息（只读）。
    public func getKeyboardLayout(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> KeyboardLayoutResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "input.getKeyboardLayout",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decodeOrThrow(resp, KeyboardLayoutResponse.decode)
    }

    // MARK: - 便捷快捷键

    /// Ctrl+C 复制。
    public func copy(pcPeerId: String, mobileDid: String? = nil) async throws -> KeyComboResponse {
        try await sendKeyCombo(pcPeerId: pcPeerId, key: "c", modifiers: [.ctrl], mobileDid: mobileDid)
    }

    /// Ctrl+V 粘贴。
    public func paste(pcPeerId: String, mobileDid: String? = nil) async throws -> KeyComboResponse {
        try await sendKeyCombo(pcPeerId: pcPeerId, key: "v", modifiers: [.ctrl], mobileDid: mobileDid)
    }

    /// Ctrl+Z 撤销。
    public func undo(pcPeerId: String, mobileDid: String? = nil) async throws -> KeyComboResponse {
        try await sendKeyCombo(pcPeerId: pcPeerId, key: "z", modifiers: [.ctrl], mobileDid: mobileDid)
    }

    /// Ctrl+A 全选。
    public func selectAll(pcPeerId: String, mobileDid: String? = nil) async throws -> KeyComboResponse {
        try await sendKeyCombo(pcPeerId: pcPeerId, key: "a", modifiers: [.ctrl], mobileDid: mobileDid)
    }

    /// Ctrl+S 保存。
    public func save(pcPeerId: String, mobileDid: String? = nil) async throws -> KeyComboResponse {
        try await sendKeyCombo(pcPeerId: pcPeerId, key: "s", modifiers: [.ctrl], mobileDid: mobileDid)
    }

    // MARK: - 内部

    private static func decodeOrThrow<T>(
        _ resp: TerminalRpcResponse,
        _ decoder: (String) throws -> T
    ) throws -> T {
        switch resp {
        case .success(_, let resultJson):
            return try decoder(resultJson)
        case .failure(let reqId, let msg):
            throw RemoteSkillError.remoteError(reqId: reqId, message: msg)
        }
    }
}
